### ISSUE TITLE
help

### ISSUE THREAD

**Issue by gedw99** [donseba/go-translator#1]
this is something I have a keen interest in.

Did you come up with a format yet ?  

My flow is:

1. Extract: Parse the golang application to extract to a text format format
2. Translate: check cache for translation of text format and if not there do a translation
3. Merge: Marge with all the other languages so all have it.

The Translation is a 2 phase where phase 1 is machine translation and phase 2 is human checking and fixing. 

See github.com/vorlif/spreak

**Comment by donseba** [comment]
I am working on something like you describe but I'm blocked by a repo that I depend on. I'll fork it and use that as a base for the time being, if I don't get a reply soon. 

It is based on gettext as it does basically check all base functionality one needs. 

Main workflow. 

In the templates you can use `tn` and `tl` functions. They will check if there is a translation key available in the main translations file. If not it will be added there. Then with poEdit or something similar you can load new keys from the main translations file, edit them, update and deploy. 

It's all I need in my current application but obviously there is room to grow.

**Comment by gedw99** [comment]
Sound awesome @donseba 

thanks for considering this btw

**Comment by gedw99** [comment]
I get the translations using google translate wrapped with golang btw .

can show you how if you want ??

**Comment by donseba** [comment]
getting the translations from google would be really cool. 

I have added the initial code and the repo i depended on is forked here : https://github.com/donseba/gotext , the build is failing since it was based on the forked repo. 

It is highly opinionated so be advised 😄  , it is also missing a lot of functionality since I havent had a use case for it yet.

**Comment by gedw99** [comment]
cheers @donseba 

I will have a look and  then add the google translate golang code.. It runs locally by calling into a browser that loads google translate  and then scrapes it. Because of the cache, it only gets used when there is a cache miss which is rare.

**Comment by gedw99** [comment]
This is one of the free google translators.

https://github.com/bas24/googletranslatefree

There are other packages out there too that do the exact same thing.

Because of the translation cache  based design we are using we will never hit its limits.  

The cache needs to be saved in git though and so it needs to be a yaml or json one.


Later if we want to get fancy, we could also easily add other paid translators too using a factory pattern later but probably over kill for now.

**Comment by gedw99** [comment]
The translation cache needs to be git merge tolerant.

Here is one but there are many out there …

https://github.com/pnkj-kmr/simple-json-db/tree/main

**Comment by donseba** [comment]
this has come a long way in a short time period. Originally, I set things up so my translation keys live in a .pot file, with a separate .po file for each language. This was all about sticking with gettext, which is the go-to for handling translations in apps—a big step up from our starting point. (atleast for me since I've worked with it for over 15 years now.)

Now, about moving to JSON for the translations, if you think you can make it work, especially with keeping up with all those tricky plural rules (you know, most languages have a couple of forms, but then you've got Arabic and Russian throwing curveballs with five or six), I say, why not give it a shot? The main thing is making sure we still play nice with the whole plural situation. also it should be configurable. 

I've already got the basics down for a tool that can handle all the add, edit, and delete stuff for the translation files. So, that's sorted and I will start development on that soon.

Next up, I'm thinking about shaking things up and maybe make a database driven version.

**Comment by gedw99** [comment]
hey @donseba 

> this has come a long way in a short time period. Originally, I set things up so my translation keys live in a .pot file, with a separate .po file for each language. This was all about sticking with gettext, which is the go-to for handling translations in apps—a big step up from our starting point. (atleast for me since I've worked with it for over 15 years now.)



> 
> Now, about moving to JSON for the translations, if you think you can make it work, especially with keeping up with all those tricky plural rules (you know, most languages have a couple of forms, but then you've got Arabic and Russian throwing curveballs with five or six), I say, why not give it a shot? The main thing is making sure we still play nice with the whole plural situation. also it should be configurable.

yes I agree that plurals a touch one. Russian and Japanese have many if I recall.

The only reason for the JSON is so that we can cache and search the cache and it can be pulled and pushed into git.  

> 
> I've already got the basics down for a tool that can handle all the add, edit, and delete stuff for the translation files. So, that's sorted and I will start development on that soon.
> 
> Next up, I'm thinking about shaking things up and maybe make a database driven version.

If you mean that you want to store the language translation inside a DB, then I strongly discourage this.  As I suggested above, the translation data, would be better in the file system so that developers can easily gen them and also pul and push them with git.

**Comment by gedw99** [comment]
I did some digging inside the Pokemon repo and think that a DB that is based on CSV is ok
https://github.com/PokeAPI/pokeapi/blob/master/data/v2/csv/translations/cs.csv

Its nice and agnostic.

so the idea is that we use JSON for the Development, and can then roll it up to CSV for Deployment. A bit like how they do it. 

then people can transact the css into any db they want if they want to go that way.

In summary we are keeping things nice and loose so that anyone can then use go-htmx with any style of development and deployment pipelines that they want to us.

Let me know if you hate it :)
